### PR TITLE
Adding a new lognormal datamodel

### LINKED ISSUE #4: Add Log-Normal data model
## Description of feature:

Add log normal data model 

## Desired behaviour:

Be able to specify a datamodel whereby log(y) ~ N(mu+log(x),sigma^2) where mu is the log of a coverage ratio

## Definition of Done (DoD):

Can use a log normal data model

## Links to related/relevant information:

[Any links to related/relevant information to assist in implementation i.e. similar methods in alternative packages]

## Package version/s (if behaviour required across multiple dpmaccount versions):

[dpmaccount versions]

### PR DESCRIPTION
**Pull request by CompBioDan**
# Requester section

*To be completed by the pull-requester (assignee)*

## Description
<details><summary>Please include a summary of the changes</summary>
 
  - What is this change?
  - What does it fix?
  - Is this a bug fix or a feature and does it break any existing functionality?
  - How has it been tested?
</details>
 
This pull request adds lognormal datamodel functionality. This includes both nopar (no scale ratio) and par (scale ratio) versions. 
 
## Type of change
- [ ] Bug fix - *non-breaking change*
- [X] New feature - *non-breaking change*
- [ ] Breaking change - *backwards incompatible change, changes expected behaviour*
- [ ] Non-user facing change, structural change, dev functionality, docs ...
 
## Requestor checklist:

*If applicable, if not indicate why in requester comments section*

- [X] I have performed a self-review of my own code
- [X] I have commented my code appropriately, focusing on explaining my design decisions (explain why, not how)
- [X] I have made corresponding changes to the documentation (comments, docstring, etc.) using roxygen (devtools check()/document() etc.)
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes
- [X] I have updated the change log/version number [CHANGELOG.md](https://github.com/ONSdigital/dpmaccount/blob/main/CHANGELOG.md)
- [X] I have checked the pipeline runs with test data

### Requester comments
 
A general guide followed to add the new datamodel (covering the new additions/modifications) is here for reference:
[dpmaccount_datamodel_guide.pdf](https://github.com/user-attachments/files/21388860/dpmaccount_datamodel_guide.pdf)

<br>
 
# Reviewer section

*To be completed by the pull-request reviewer*

##  Peer review checklist:
Any new code (where appropriate) includes all of the following:

- [ ] **Commit messages**:
  - [ ] The individual commit messages are sufficiently descriptive, when combined with the **pull request description** above, forming a coherent audit trail. If not, add suggested improvements/changes to the reviewer notes below
  - [ ] The commit messages are in-line with the [contributing guidance](https://github.com/ONSdigital/dpmaccount/blob/main/CONTRIBUTING.md)
- [ ] **Data**: All [personally identifiable data](https://ico.org.uk/for-organisations/uk-gdpr-guidance-and-resources/personal-information-what-is-it/what-is-personal-information-a-guide/) has been removed/anonymised
- [ ] **Documentation**: Docstrings, comments have been added/updated
- [ ] **[Style guidelines](https://confluence.ons.gov.uk/display/PP/Style+guidelines)**: New code conforms to the project's contribution guidelines
- [ ] **[Functionality](https://confluence.ons.gov.uk/display/PP/Functionality)**: The code works fully implements the requirements and works as expected, handles expected edge cases, exceptions are handled appropriately
- [ ] **[Complexity](https://confluence.ons.gov.uk/display/PP/Complexity)**: The code is not overly complex, logic has been split into appropriately sized functions, etc.
- [ ] **[Test coverage](https://confluence.ons.gov.uk/display/PP/Test+coverage)**: Unit tests cover essential functions for a reasonable range of inputs and conditions. Added and existing tests pass on my machine.
- [ ] **Vignettes**: Any relevant vignettes have been modified/created.
- [ ] **Version numbering**: The versioning number has been incremented correctly and follows mandated semantic versioning protocol in [contributing guidance](https://github.com/ONSdigital/dpmaccount/blob/main/CONTRIBUTING.md).

*If any of the above are not applicable, include an explanation in the review comments below*

### Review comments
Suggestions should be tailored to the code that you are reviewing. Provide context.
Be critical and clear, but not mean. Ask questions and set actions.
<details><summary>These might include:</summary>
 
- bugs that need fixing (does it work as expected? and does it work with other code that it is likely to interact with?)
- alternative methods (could it be written more efficiently or with more clarity?)
- documentation improvements (does the documentation reflect how the code actually works?)
- additional tests that should be implemented
- Do the tests effectively assure that it works correctly? Are there additional edge cases/negative tests to be considered?
- Code style improvements (could the code be written more clearly?)
</details>
<br>
 
*Further reading: [code review best practices](https://best-practice-and-impact.github.io/qa-of-code-guidance/peer_review.html)*

### CODE REVIEW

File: R/datamod-constructors.R [hunk 1]
```diff
@@ -1177,3 +1177,280 @@ new_datamod_poisson <- function(data,
...
+#' - C are values of i for a particular combination of cohort, sex, and region
+#' - \eqn{x_i} is the true (unobserved) count for cell \eqn{i}
+#' - \eqn{y_i} is the observed count for cell \eqn{i}
+#' - \eqn{\mu} is the net coverage ratio for cell \eqn{i}
+#' - \eqn{\sigma_i} is the standard deviation for cell \eqn{i}
+#'
+#' When `scale_ratio` is greater than `0`,
+#' the data model becomes
+#'
+#' \deqn{l(x_i \mid y_i, \mu_i^{(s)}, \sigma^{(s)}) = -\sum_{i \in C} \frac{(log(y_i) - log(x_i) - \mu_i ^{(s)} - {\alpha_{c_i}})^2}{2 (\sigma^{(s)})^2})}
```
Comment (tayloa1) [inline R/datamod-constructors.R:1209]: Possible extra { - need to check rendering.

File: R/datamod-constructors.R [hunk 2]
```diff
@@ -1177,3 +1177,280 @@ new_datamod_poisson <- function(data,
...
+#' then its coverage ratio is 1.
+#'
+#' `ratio` can be a single number or a data frame.
+#' If `ratio` is a single number, then the
+#' same coverage ratio is used for all values of
+#' the data. If `ratio` is a data frame,
+#' then different coverage rates are used for
+#' different parts of the data.
+#' If a classifying variable in `data` is
+#' omitted from `ratio`,  then the coverage ratio
```
Comment (tayloa1) [inline R/datamod-constructors.R:1258]: Double space between comma and 'then'

File: R/datamod-constructors.R [hunk 3]
```diff
@@ -1177,3 +1177,280 @@ new_datamod_poisson <- function(data,
...
+#'
+#' When `scale_ratio` and `scale_sd` are both `0`
+#' (the default), the log-normal data model has a
+#' probability density function
+#'
+#' \deqn{f_x(x) =  \frac{1}{x \sigma \sqrt{2 \pi}} e^{-\frac{(ln x - \mu)^2}{2 \sigma^2}}}
+#'
+#' which results in a log-posterior contribution of
+#'
+#' \deqn{l(x_i \mid y_i, \mu_i^{(s)}, \sigma^{(s)}) = -\sum_{i \in C} \frac{(log(y_i) - log(x_i) - \mu_i ^{(s)})^2}{2 (\sigma^{(s)})^2})}
```
Comment (tayloa1) [inline R/datamod-constructors.R:1196]: Additional ')' in line 1196 [second character from end of string]

Should instead be 
#' \deqn{l(x_i \mid y_i, \mu_i^{(s)}, \sigma^{(s)}) = -\sum_{i \in C} \frac{(log(y_i) - log(x_i) - \mu_i ^{(s)})^2}{2 (\sigma^{(s)})^2}}

Also 's' isn't defined - it may not be obvious what this represents. 

Maybe add a line below line 1204 that says something like
\eqn{s} is the s-th sample drawn from the posterior distribution during estimation.

File: R/datamod-constructors.R (hunk 1 above)
Comment (tayloa1) [inline R/datamod-constructors.R:1209]: As above (line 1196) additional ')' - second character from end needs removing

**Review by tayloa1 — CHANGES_REQUESTED** [review]
All test have passed.
Package can be built and installed as expected.

Some very minor changes required in documentation (extra brackets, a double space and a definition for an undefined term)
Also there is duplication of the get_nms_par method for poisson_nopar in codatamod-methods.R which should be removed.

File: R/datamod-constructors.R (hunk 3 above)
[the remark this answers is not in the document]
Comment (CompBioDan) [reply R/datamod-constructors.R:1196]: I think you're right, there are 9 brackets opened and 10 brackets closed.

Comment (CompBioDan) [reply R/datamod-constructors.R]: Removed the erroneous ')' and adding the definition for 's'.

File: R/datamod-constructors.R (hunk 2 above)
Comment (CompBioDan) [reply R/datamod-constructors.R]: Removed the double space between he comma and 'then'.

File: R/datamod-constructors.R (hunk 1 above)
Comment (tayloa1) [reply R/datamod-constructors.R]: Yes, you're right I should have written ')' not '{' but can see this has been removed.

**Review by tayloa1 — CHANGES_REQUESTED** [review]
All changes look good. 

However, R-CMD-check failing for macos is failing. And all 3 OS checks aren't reporting their status to the PR, so I couldn't approve without over-riding. I've tried re-running the action (because I don't recall seeing this before and it's not obvious why there should be a problem now given you've not changed anything here) but got the same result. (maybe due to an update to the latest mac os?)

See here for details https://github.com/ONSdigital/dpmaccount/actions/runs/17378826651/job/49487380238

But it seems to boil down to this line:

dpmaccount.cpp:428:28: error: call to 'isfinite' is ambiguous
  428 |   int sd_stk_init_finite = isfinite(sd_stk_init);
  
 Potentially looks like this could be resolved by replacing the above in the cpp code with 
 int sd_stk_init_finite = std::isfinite(sd_stk_init);
 
 Edit: had a crisis of confidence and googled to check that would be sufficient... it's not. You'd also need to add 
 
` #include <cmath>`

to the top of the cpp code.

**Comment by CompBioDan** [comment]
It looks like the R-CMD-check error for macOS is being caused by the recent change to use macOS-15 as the macos-latest in the [R-CMD-check.yaml](https://github.com/ONSdigital/dpmaccount/blob/main/.github/workflows/R-CMD-check.yaml). Changing this to macos-14 allows the macOS checks to pass, suggesting we need to find a permanent fix for macOS-15 (likely working with John who is the current sole developer on macOS). In the meantime I don't believe we will be able to disable the macos-latest check without modifying the .yaml in main which can only be done through merging. 

Reference to macos-latest update:
**The macos-latest label will migrate to macOS 15 beginning August 4, 2025. For more information see https://github.com/actions/runner-images/issues/12520**

**Review by tayloa1 — APPROVED** [review]
Most recent commit fixes R-CMD-check failure for macOS. Change was previously verified to prevent compiler failure on an installation of macOS as per issue #5.

Recent change introduces no new errors, all tests pass, package installs locally and pipeline code (with appropriate changes that require dpmaccount v0.3.1) runs successfully.
